pinot-server/src/main/java/org/apache/pinot/server/api/resources/ReIngestionResource.java

```diff
@@ -0,0 +1,361 @@
+    if (instanceDataManager == null) {
+      throw new WebApplicationException("Invalid server initialization", Response.Status.INTERNAL_SERVER_ERROR);
+    }
+
+    RealtimeTableDataManager tableDataManager =
+        (RealtimeTableDataManager) instanceDataManager.getTableDataManager(tableNameWithType);
+    if (tableDataManager == null) {
+      throw new WebApplicationException("Table data manager not found for table: " + tableNameWithType,
+          Response.Status.NOT_FOUND);
+    }
+
+    IndexLoadingConfig indexLoadingConfig = tableDataManager.fetchIndexLoadingConfig();
+    LOGGER.info("Executing re-ingestion for table: {}, segment: {}", tableNameWithType, segmentName);
+
+    // Get TableConfig, Schema, ZK metadata
+    TableConfig tableConfig = indexLoadingConfig.getTableConfig();
+    if (tableConfig == null) {
+      throw new WebApplicationException("Table config not found for table: " + tableNameWithType,
+          Response.Status.NOT_FOUND);
+    }
+    Schema schema = indexLoadingConfig.getSchema();
+    if (schema == null) {
+      throw new WebApplicationException("Schema not found for table: " + tableNameWithType,
+          Response.Status.NOT_FOUND);
+    }
+
+    SegmentZKMetadata segmentZKMetadata = tableDataManager.fetchZKMetadata(segmentName);
+    if (segmentZKMetadata == null) {
+      throw new WebApplicationException("Segment metadata not found for segment: " + segmentName,
+          Response.Status.NOT_FOUND);
+    }
+
+    // Check if download url is present
+    if (segmentZKMetadata.getDownloadUrl() != null) {
+      throw new WebApplicationException(
+          "Download URL is already present for segment: " + segmentName + ". No need to re-ingest.",
+          Response.Status.BAD_REQUEST);
+    }
+
+    // Grab start/end offsets
+    String startOffsetStr = segmentZKMetadata.getStartOffset();
+    String endOffsetStr = segmentZKMetadata.getEndOffset();
+    if (startOffsetStr == null || endOffsetStr == null) {
+      throw new WebApplicationException("Null start/end offset for segment: " + segmentName,
+          Response.Status.INTERNAL_SERVER_ERROR);
+    }
+
+    // Check if this segment is already being re-ingested
+    AtomicBoolean isIngesting = SEGMENT_INGESTION_MAP.computeIfAbsent(segmentName, k -> new AtomicBoolean(false));
+    if (!isIngesting.compareAndSet(false, true)) {
+      return Response.status(Response.Status.CONFLICT)
+          .entity("Re-ingestion for segment: " + segmentName + " is already in progress.")
+          .build();
+    }
+
+    // Generate a jobId for tracking
+    String jobId = UUID.randomUUID().toString();
+    ReIngestionJob job = new ReIngestionJob(jobId, tableNameWithType, segmentName);
+
+    // Kick off the actual work asynchronously
+    REINGESTION_EXECUTOR.submit(() -> {
+      try {
+        int partitionGroupId = llcSegmentName.getPartitionGroupId();
+
+        Map<String, String> streamConfigMap = IngestionConfigUtils.getStreamConfigMaps(tableConfig).get(0);
```

Comment:
Is there a way to know which stream config to use to generate the segment?

Reply:
Not possible. Ideally though in multi-stream only broker and topic names should differ across stream configs, rest everything should remain the same